onmt/utils/distributed.py

```diff
@@ -176,3 +181,110 @@ def consumer(process_fn, opt, device_id, error_queue):  # noqa: E501
         import traceback
 
         error_queue.put((opt.gpu_ranks[device_id], traceback.format_exc()))
+
+
+def spawned_infer(opt, device_id, error_queue, queue_instruct, queue_result):
+    """Run `process_fn` on `device_id` with data from `batch_queue`."""
+    try:
+        gpu_rank = multi_init(opt, device_id)
+        if gpu_rank != opt.gpu_ranks[device_id]:
+            raise AssertionError(
+                "An error occurred in \
+                  Distributed initialization"
+            )
+        torch.cuda.set_device(device_id)
+        init_logger(opt.log_file)
+        translator = build_translator(opt, device_id, logger=logger, report_score=True)
+        transforms_cls = get_transforms_cls(opt._all_transform)
+        while True:
+            instruction = queue_instruct.get()
+            if instruction[0] == "stop":
+                break
+            elif instruction[0] == "infer_list":
+                src = instruction[1]
+                infer_iter = build_dynamic_dataset_iter(
+                    opt,
+                    transforms_cls,
+                    translator.vocabs,
+                    task=CorpusTask.INFER,
+                    src=src,
+                )
+                infer_iter = IterOnDevice(infer_iter, device_id)
+                scores, preds = translator._translate(
+                    infer_iter, infer_iter.transform, opt.attn_debug, opt.align_debug
+                )
+                queue_result.put(scores)
+                queue_result.put(preds)
+            elif instruction[0] == "infer_file":
+                src = instruction[1]
+                infer_iter = build_dynamic_dataset_iter(
+                    opt, transforms_cls, translator.vocabs, task=CorpusTask.INFER
+                )
+                infer_iter = IterOnDevice(infer_iter, device_id)
+                scores, preds = translator._translate(
+                    infer_iter, infer_iter.transform, opt.attn_debug, opt.align_debug
+                )
+                queue_result.put(scores)
+                queue_result.put(preds)
+
+    except KeyboardInterrupt:
+        pass  # killed by parent, do nothing
+    except Exception:
+        # propagate exception to parent process, keeping original traceback
+        import traceback
+
+        error_queue.put((opt.gpu_ranks[device_id], traceback.format_exc()))
+
+
+class mp_inference(object):
+    def __init__(self, opt):
```

Comment:
Maybe add some docstrings. Also weird to have the classname in snake_case. `MPInference` instead?